Modified pdx to disassemble hexpairs

tantei3 opened the pull request:
Closes #14771 

Modified the `pdx` command to disassemble the hexpairs by aliasing it to `pd @x:<hexpairs>` and get the output as mentioned in the issue. Took the reference from `tpx` command for implementation.

radare commented:
i did that already last week fforgot to close the issue :(

but in your pr you are removing my code and redoing it in a different way

tantei3 commented:
![image](https://user-images.githubusercontent.com/54449449/63650288-a4fac400-c766-11e9-90e7-528fd7bc999b.png)

Currently for the hex size bigger than like 4 bytes, it doesn't display the disassembly as expected, but this PR takes care of moving the logic above the block size check. Also the issue mentioned about aliasing to `pd @x:` so I thought that this change was required!

radare commented:
Hexpairs dont start with 0x

Syntactically your example is wrong because it must be assuming endianness. And this should work fine if you remove the 0x of the begining

radare commented:
If broken pls fix and we probably need some tests

tantei3 commented:
![image](https://user-images.githubusercontent.com/54449449/63651346-5868b600-c771-11e9-9be7-f27f169dbc32.png)

Well looking at these examples, I don't see any differences even if 0x is included or not (I even tried with this `31c09952682f636174682f62696e89e3526873737764682f2f7061682f65746389e1b00b52515389e1cd80` as the input)

I don't know what is the expected behaviour, and how to add test cases for example.

radare commented:
Then thats a bug in my impl :)

> On 25 Aug 2019, at 16:22, tantei3 <notifications@github.com> wrote:
> 
> 
> Well looking at these examples, I don't see any differences even if 0x is included or not (I even tried with this 31c09952682f636174682f62696e89e3526873737764682f2f7061682f65746389e1b00b52515389e1cd80 as the input)
> 
> I don't know what is the expected behaviour, and how to add test cases for example.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

radare commented on libr/core/cmd_print.c:
```diff
@@ -5330,7 +5330,18 @@ static int cmd_print(void *data, const char *input) {
+				r_cons_printf ("Usage: pdx [hexpairs] - disassemble hexpairs\n");
+				return 0;
+			} else {
+				r_core_cmdf (core, "pd @x:%s", r_str_trim_ro (input + 2));
```
this can result in command injection issues, will be better to use the API to do that.

tantei3 replied:
Okay so I should call `__cmd_pad` function that is correctly implemented instead of the `r_core_cmdf` function before the block computation is resolved for the `pd` command, right? 

What about the `tpx` command? That also does a similar string formatting, wouldn't it be affected by the same command injection?

radare commented on libr/core/cmd_print.c:
```diff
@@ -5330,7 +5330,18 @@ static int cmd_print(void *data, const char *input) {
+				return 0;
+			} else {
+				r_core_cmdf (core, "pd @x:%s", r_str_trim_ro (input + 2));
+				eprintf ("pd @x:%s\n", r_str_trim_ro (input + 2));
```
this seems like a debug print leftover

radare replied:
remove this line

radare commented:
Can you add tests and address the comments? i can confirm the issue you mention

tantei3 commented:
Can I get some links and resources on how to proceed with adding test cases? It would be helpful to know which files to edit in the repository and which files to look at as reference as well, since I am first time looking into this!

radare commented:
see other files in the radare2-regressions repository (under new/ db/*/*). to run then node_modules/.bin/r2r -i db/path/test

radare commented:
i think in the book there are some pages about this

radare commented on libr/core/cmd_print.c:
```diff
@@ -5330,7 +5330,18 @@ static int cmd_print(void *data, const char *input) {
+				r_cons_printf ("Usage: pdx [hexpairs] - disassemble hexpairs\n");
+				return 0;
+			} else {
+				r_core_cmdf (core, "pd @x:%s", r_str_trim_ro (input + 2));
```
probably better to parse the hexpair string and return it back into hexpairs from the raw buffer. this way it wont be possible to inject code

radare commented:
ping

radare commented:
adding tests is easy, if you need any help drop a msg on the irc/telegram channel, but dont hold this issue for too much, can you address the comments quickly? i plan to release this sunday

> On 26 Aug 2019, at 18:43, tantei3 <notifications@github.com> wrote:
> 
> Can I get some links and resources on how to proceed with adding test cases? It would be helpful to know which files to edit in the repository and which files to look at as reference as well, since I am first time looking into this!
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/14892?email_source=notifications&email_token=AAG75FUROK37NONMYZDVEJLQGQB4BA5CNFSM4IPILEA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5E53KY#issuecomment-524934571>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAG75FXC4GNP4KKN4UOHN53QGQB4BANCNFSM4IPILEAQ>.
>

tantei3 commented:
Sorry for the delay, I added a few basic test cases for `pad`, `pix` and `pdx`. Also noticed that the current issue is that for the hexpairs that are in the range of UINT64 get compared to block size and hence that error is thrown for pix and pdx in some cases. To fix this, I have moved the evaluation of `pdx` above the compare with blocksize part, and for the `pix` case, I have added the bypass for the `i` commands.

radare commented:
i have fixed the conflict and the commit message and merged. hopefully will be green , otherwise ill chk whats wrong and fix it

radare commented:
thanks!
